Catch unsupported model types in `BinaryThresholdPredictor`.

ablaom opened the pull request:
To resolve #486.

Also adds a docstring to resolve https://github.com/alan-turing-institute/MLJ.jl/issues/973.

ablaom commented:
@OkonSamuel Can you please review this?

OkonSamuel commented on src/builtins/ThresholdPredictors.jl:
```diff
@@ -67,24 +67,96 @@ const ERR_CLASSES_DETECTOR = ArgumentError(
+Xbalanced, ybalanced = make_moons(10000, xshift=0.5,noise=5.0, rng=rng)
+ybalanced = coerce(ybalanced, OrderedFactor)
+mask = map(ybalanced) do class
+    class == 0 || rand() < 0.1
```
```suggestion
    class == 0 || rand(rng) < 0.1
```

ablaom replied:
Good spot.

OkonSamuel commented on src/builtins/ThresholdPredictors.jl:
````diff
@@ -67,24 +67,96 @@ const ERR_CLASSES_DETECTOR = ArgumentError(
+Estimating the performance of the auto-thresholding model:
+
+```julia
+e = evaluate!(mach, resampling=CV(nfolds=6), measure=[f1score, accuracy])
````
Don't we need to create a new machine 
`mach2 = machine(report(mach).best_model, X, y)`
before evaluating it. Else the tuning process will be repeated for each fold of the evaluation process.

ablaom replied:
> Else the tuning process will be repeated for each fold of the evaluation process. This would mean that the threshold value used in every fold of the evaluation would be different.

Yes, this is exactly what we want! Our wrapper model is "self-tuning". Meaning, every time you fit it, it learns the best threshold. See https://mlr.mlr-org.com/articles/tutorial/nested_resampling.html .

OkonSamuel commented on src/builtins/ThresholdPredictors.jl:
```diff
@@ -67,24 +67,96 @@ const ERR_CLASSES_DETECTOR = ArgumentError(
+    n=30,
+)
+mach = machine(tuned_point_predictor, X, y) |> fit!
+report(mach).best_model.threshold # 0.813
```
```suggestion
report(mach).best_model.threshold # 0.102
```

OkonSamuel commented on src/builtins/ThresholdPredictors.jl:
````diff
@@ -67,24 +67,96 @@ const ERR_CLASSES_DETECTOR = ArgumentError(
+
+```julia
+e = evaluate!(mach, resampling=CV(nfolds=6), measure=[f1score, accuracy])
+e.measurement[1] # 0.0576
````
This would also need to be updated.

OkonSamuel commented on src/builtins/ThresholdPredictors.jl:
```diff
@@ -101,8 +173,10 @@ function BinaryThresholdPredictor(args...;
-    metamodel =
-        _type_given_atom[MMI.abstract_type(atom)](atom, Float64(threshold))
+    A = MMI.abstract_type(atom)
+    T = get(_type_given_atom, A, nothing)
+    isnothing(T) && throw(err_unsupported_model_type(A))
```
```suggestion
   T = threshold_constructor(A)
```

Then somewhere below/above add the following
```julia
for (atom, type) in type_given_atom
    @eval threshold_constructor(::Type{$atom}) = $type
end
threshold_constructor(A) = throw(err_unsupported_model_type(A))

This approach is type stable.

ablaom replied:
Done.

OkonSamuel reviewed (changes requested):
Hello Anthony,

I'm sorry for the late review, I thought I had already given my review. lol
Anyways, I left some comments.

ablaom commented:
I have replaced the synthetic data with the Pima Indian diabetes data, and replaced f1score with balanced_accuracy. In this case one sees a "statistically significant" improvement in the objective function.
